### PR TITLE
Fix HDF5Writer id()-reuse bug corrupting class_index in multi-worker generation

### LINKED ISSUE #478: HDF5Writer: id()-reuse bug corrupts class_index (and other metadata) in multi-worker dataset generation
## Summary

`HDF5Writer` (`torchsig/utils/file_handlers/hdf5.py`) can silently corrupt metadata for a fraction of samples in any dataset generated via `DatasetCreator` with `num_workers > 1` (i.e. any real-scale generation using `WorkerSeedingDataLoader`). The most visible symptom: reading the dataset back with `StaticTorchSigDataset` occasionally returns `class_index` (or other target labels) as a **list** instead of a scalar, containing values from an unrelated sample.

Observed corruption rate scaled with dataset size in our testing: 0% in a single-process (`num_workers=0`) generation, ~1.4% at 3,000 samples / 4 workers, and **~67%** on a real full-scale 1,060,000-sample generation (10 workers). It reproduces at `Impairments(level=0)` ("clean") with no OFDM signals involved, so it is not specific to any signal type or impairment level — any multi-worker generation is at risk, and the corruption rate grows with run size/duration.

## Root cause

`_hdf5_key()` returns a stable key for an object if it has a `._hdf5_key` attribute, otherwise falls back to `str(id(obj))`:

```python
def _hdf5_key(obj) -> str:
    try:
        return obj._hdf5_key
    except AttributeError:
        return str(id(obj))
```

`HDF5Writer._assign_hdf5_keys()` stamps a monotonically-increasing, collision-free `_hdf5_key` on each `Signal` object (and its `component_signals`) immediately before writing a batch - but it does **not** stamp the signal's metadata `.parent` chain (the generator/class-level `HierarchicalMetadataObject` that `populate_hdf5_group_with_metadata()` recurses into). Those parent objects therefore fall back to `str(id(obj))`.

Parent metadata objects are not guaranteed to stay alive for the whole write session (e.g. a new one can be constructed per batch/worker and then garbage-collected once its signals are written). When CPython recycles a freed parent's memory address for a later, unrelated parent object, both resolve to the **same** `str(id(...))` key. `populate_hdf5_group_with_metadata()`'s dedup guard then treats the second (different!) parent as already-written and skips it:

```python
def populate_hdf5_group_with_metadata(group, metadata_obj) -> bool:
    key = _hdf5_key(metadata_obj)
    if key in group:
        return False   # <-- silently keeps the FIRST parent's data
    ...
```

Every signal whose `.parent` pointed at the second (skipped) object now reads back the first, unrelated parent's metadata fields on load - including `class_index`, which is why `StaticTorchSigDataset.__getitem__` sometimes returns a list like `[correct_label, wrong_label]` (the recursive fetch pulling from both the signal's own metadata and a mismatched/aliased parent) instead of a clean scalar.

## Reproduction

Using `TorchSigIterableDataset` + `WorkerSeedingDataLoader` + `DatasetCreator` (the standard pattern from `scripts/generate_dataset_from_config.py`/the README), generate a dataset with `num_workers > 1`, then read it back and check `class_index`:

```python
from torchsig.datasets.datasets import TorchSigIterableDataset, StaticTorchSigDataset
from torchsig.transforms.impairments import Impairments
from torchsig.utils.data_loading import WorkerSeedingDataLoader
from torchsig.utils.defaults import TorchSigDefaults
from torchsig.utils.writer import DatasetCreator, identity_collate_fn

metadata = TorchSigDefaults().default_dataset_metadata
metadata["num_iq_samples_dataset"] = 4096
metadata["num_signals_min"] = 1
metadata["num_signals_max"] = 1
metadata["signal_duration_in_samples_min"] = 4096 * 0.8
metadata["signal_duration_in_samples_max"] = 4096 * 1.0

dataset = TorchSigIterableDataset(
    metadata=metadata,
    transforms=[Impairments(level=0).dataset_transforms],
    component_transforms=[Impairments(level=0).signal_transforms],
    target_labels=["class_index"],
    seed=1234567890,
)
dataloader = WorkerSeedingDataLoader(
    dataset, seed=1234567890, batch_size=16, num_workers=10, collate_fn=identity_collate_fn,
)
creator = DatasetCreator(
    dataloader=dataloader, dataset_length=50000, root="repro_out", overwrite=True, multithreading=True,
)
creator.create()

ds = StaticTorchSigDataset(root="repro_out", target_labels=["class_index"])
bad = [i for i in range(len(ds)) if isinstance(ds[i][1], list)]
print(f"bad: {len(bad)}/{len(ds)}")   # observed non-zero at this scale/worker count
```

In our runs this produced 0/3000 at `num_workers=0`, ~42/3000 at `num_workers=4`, and 0/50000 *after* the fix below, at `num_workers=10` (pre-fix this scale reliably reproduced the bug too).

## Suggested fix

Stamp a stable, counter-based key (same mechanism already used for `Signal` objects) on the metadata `.parent` chain too, reusing the existing key via `hasattr` when the same parent object is genuinely shared across many signals (so intentional dedup still works):

```python
def _assign_hdf5_keys(self, signal) -> None:
    signal._hdf5_key = str(self._key_counter)
    self._key_counter += 1
    self._assign_hdf5_keys_to_parent_chain(signal)
    for cs in signal.component_signals:
        self._assign_hdf5_keys(cs)

def _assign_hdf5_keys_to_parent_chain(self, metadata_obj) -> None:
    parent = getattr(metadata_obj, "parent", None)
    while parent is not None:
        if not hasattr(parent, "_hdf5_key"):
            parent._hdf5_key = str(self._key_counter)
            self._key_counter += 1
        parent = getattr(parent, "parent", None)
```

We applied exactly this patch locally and confirmed it resolves the issue: the same 50,000-sample/10-worker case that reproduced corruption pre-fix now shows 0/50000 bad, and re-running our full 1,060,000-sample production dataset went from ~67% corrupted `class_index` values to 0%.

Happy to open a PR with this change (plus a regression test) if that's useful - let me know.

## Environment

- torchsig 2.1.1 (installed via `pip install -e .` from a clone at commit `d9abfe1`)
- Python via conda env, Linux

### PR DESCRIPTION
**Pull request by sclkim1**
Fixes #478.

## Problem

`HDF5Writer` stamps a stable, counter-based `_hdf5_key` on each `Signal` object before writing (`_assign_hdf5_keys`), but not on the signal's metadata `.parent` chain (the generator/class-level `HierarchicalMetadataObject` that `populate_hdf5_group_with_metadata()` recurses into). Those parent objects fell back to `str(id(obj))` in `_hdf5_key()`.

Parent metadata objects aren't guaranteed to stay alive for the whole write session - a new one can be built per batch/worker and then garbage-collected once its signals are written. When CPython recycles a freed parent's memory address for a later, unrelated parent object, both resolve to the same key, and `populate_hdf5_group_with_metadata()`'s dedup guard (`if key in group: return False`) silently keeps the first parent's data and skips writing the second. Every signal pointing at the second (skipped) parent then reads back the first, unrelated parent's metadata fields on load - most visibly, `class_index` coming back as a list like `[correct_label, wrong_label]` instead of a scalar.

This only shows up with `num_workers > 1` (real-scale generation via `WorkerSeedingDataLoader`), and the corruption rate grows with run size: 0% single-process, ~1.4% at 3,000 samples/4 workers, ~67% on a real 1,060,000-sample production run in our case. It is not specific to any signal type or impairment level.

See #478 for the full writeup.

## Fix

`HDF5Writer._assign_hdf5_keys()` now also stamps a stable key on the metadata `.parent` chain (`_assign_hdf5_keys_to_parent_chain`), reusing an existing key via `hasattr` so a genuinely-shared parent object still dedupes to one HDF5 group - only objects without a key yet (i.e. those at risk of an id() collision) get a fresh counter-based one.

## Testing

- Added `test_writer_no_parent_metadata_id_reuse_corruption` (`tests/utils/test_writer.py`, marked `@pytest.mark.full`): generates 3,000 samples through the real multi-worker `WorkerSeedingDataLoader`/`DatasetCreator` path and asserts no sample's `class_index` comes back as a list. Passes locally after the fix.
- Manually reran a 50,000-sample/10-worker generation that reliably reproduced the bug pre-fix; 0/50000 corrupted afterward.
- Regenerated a real 1,060,000-sample + 106,000-sample production dataset with the fix; corruption went from ~67% to 0/1,166,000.

### CODE REVIEW

**Comment by sclkim1** [comment]
Closing this PR - not looking to carry changes in TorchDSP/torchsig right now. The underlying bug report stays open at #478 in case it's useful; the fix is available there if anyone wants to pick it up.
